Create notebook for introducing workspace resources

emmarogge opened the pull request:
See [PR 31](https://github.com/DataBiosphere/terra-axon-examples/pull/31) for additional context. Due to a rogue commit, I had to create a new PR for the changes to this file.

emmarogge committed:
Create notebook for introducing workspace resources

See [PR 31](https://github.com/DataBiosphere/terra-axon-examples/pull/31) for additional context. Due to a rogue commit, I had to create a new PR for the changes to this file.

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "Widget input parameters include:\n",
+    "\n",
+    "- `Access`: Defaults to `SHARED_ACCESS`, which grants bucket access to all users with access to the workspace, based on the user's workspace access (Reader or Writer). `PRIVATE_ACCESS` ensures the bucket and its contents cannot be shared.\n",
+    "- `Auto-delete`: Number of days after which to auto-delete the objects in the bucket. This option is a shortcut for specifying a [lifecycle rule](https://github.com/DataBiosphere/terra-cli#gcs-bucket-lifecycle-rules). If no value is provided,\n",
```
Looks like you need to complete this sentence

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "- `Cloning`: Read about these options [in the Cloning Instruction section](#choosing-the-right-cloning-instruction).\n",
+    "- `Name`: The name by which you will reference this resource in your workspace. A short and descriptive name is suggested, e.g., `human-genomic-data-table`.\n",
+    "- `Description`: A description of this resource's contents and purpose.\n",
+    "- `Project ID`: The [Google Cloud project ID](https://cloud.google.com/docs/overview#projects) for your table.\n",
+    "- `Dataset ID`: The [BigQuery dataset ID](https://google.com/bigquery/docs/datasets-intro#datasets) for your table.\n",
```
```suggestion
    "- `Project ID`: The [Google Cloud project ID](https://cloud.google.com/docs/overview#projects) of your table.\n",
    "- `Dataset ID`: The [BigQuery dataset ID](https://google.com/bigquery/docs/datasets-intro#datasets) of your table.\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "Then, populate the input fields:\n",
+    "\n",
+    "- `Cloning`: Read about these options [in the Cloning Instruction section](#choosing-the-right-cloning-instruction).\n",
+    "- `Project ID`: The Google Cloud Project ID for your datset.\n",
+    "- `Dataset ID`: The BigQuery dataset ID for your dataset.\n",
```
```suggestion
    "- `Project ID`: The Google Cloud Project ID of your datset.\n",
    "- `Dataset ID`: The BigQuery dataset ID of your dataset.\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "- `Path`: Consists of the project ID and dataset ID with format `<PROJECT_ID>.<DATASET_ID>`.\n",
+    "- `Name`: The name by which you will reference this resource in your workspace. A short and descriptive name is suggested, e.g., `human-genomic-data`.\n",
+    "- `Description`: A description of this resource's contents and purpose.\n",
+    "  **NOTE:** You must populate either the path field (which will have format `<PROJECT_ID>.<DATASET_ID>`) field or the both [dataset ID and the project ID](https://cloud.google.com/bigquery/docs/datasets-intro#datasets) fields for this widget to run successfully.\n",
```
```suggestion
    "  **NOTE:** You must populate either the path field (which will have format `<PROJECT_ID>.<DATASET_ID>`) or both the [dataset ID and the project ID](https://cloud.google.com/bigquery/docs/datasets-intro#datasets) fields for this widget to run successfully.\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "- `Dataset ID`: The [BigQuery dataset ID](https://google.com/bigquery/docs/datasets-intro#datasets) for your table.\n",
+    "- `Table ID`: The [BigQuery table ID](https://cloud.google.com/bigquery/docs/datasets-intro#datasets) of your table.\n",
+    "- `Path`: Consists of the project ID, dataset ID, & table ID with format `<PROJECT_ID>.<DATASET_ID>.<TABLE_ID>`.\n",
+    "  **NOTE:** You must populate either the path path field (which will have format `<PROJECT_ID>.<DATASET_ID>.<TABLE_ID>`) field, or all of the dataset ID, project ID and table ID fields.\n",
```
```suggestion
    "  **NOTE:** You must populate either the path field (which will have format `<PROJECT_ID>.<DATASET_ID>.<TABLE_ID>`), or all of the dataset ID, project ID and table ID fields.\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "\n",
+    "Widget input parameters include:\n",
+    "\n",
+    "- `Cloning`: _You cannot change the workspace ID after workspace creation._\n",
```
Please fix this line

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "Widget input parameters include:\n",
+    "\n",
+    "- `Cloning`: _You cannot change the workspace ID after workspace creation._\n",
+    "- `Bucket Name`: Must be a string that [meets requirements](https://cloud.google.com/storage/docs/buckets#naming), and must be globally unique. We suggest prepending the `<PROJECT_ID>` of the GCP project associated with your workspace, which can be found on the to a descriptive string (e.g., `<PROJECT_ID>_output_data`).\n",
```
Grammar at the end of this sentence?

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "\n",
+    "- `Cloning`: _You cannot change the workspace ID after workspace creation._\n",
+    "- `Bucket Name`: Must be a string that [meets requirements](https://cloud.google.com/storage/docs/buckets#naming), and must be globally unique. We suggest prepending the `<PROJECT_ID>` of the GCP project associated with your workspace, which can be found on the to a descriptive string (e.g., `<PROJECT_ID>_output_data`).\n",
+    "- `Description`: This is description shown in the Resources tab of your workspace's page in Verily Workbench and when the `terra resource list` command is run. The description you provide should add helpful context about the purpose and/or contents of your bucket.\n",
```
```suggestion
    "- `Description`: This is the description shown in the Resources tab of your workspace's page in Verily Workbench and when the `terra resource list` command is run. The description you provide should add helpful context about the purpose and/or contents of your bucket.\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "- `Cloning`: _You cannot change the workspace ID after workspace creation._\n",
+    "- `Bucket Name`: Must be a string that [meets requirements](https://cloud.google.com/storage/docs/buckets#naming), and must be globally unique. We suggest prepending the `<PROJECT_ID>` of the GCP project associated with your workspace, which can be found on the to a descriptive string (e.g., `<PROJECT_ID>_output_data`).\n",
+    "- `Description`: This is description shown in the Resources tab of your workspace's page in Verily Workbench and when the `terra resource list` command is run. The description you provide should add helpful context about the purpose and/or contents of your bucket.\n",
+    "- `Name`: This is the name which can be referenced in Workbench CLI commands so the value should be brief and memorable while communicating the purpose and/or contents of the bucket (e.g. `<STUDY_NAME>-data-bucket`).\n",
```
```suggestion
    "- `Name`: This is the name which can be referenced in Workbench CLI commands. The value should be brief and memorable while communicating the purpose and/or contents of the bucket (e.g. `<STUDY_NAME>-data-bucket`).\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "- `Description`: This is description shown in the Resources tab of your workspace's page in Verily Workbench and when the `terra resource list` command is run. The description you provide should add helpful context about the purpose and/or contents of your bucket.\n",
+    "- `Name`: This is the name which can be referenced in Workbench CLI commands so the value should be brief and memorable while communicating the purpose and/or contents of the bucket (e.g. `<STUDY_NAME>-data-bucket`).\n",
+    "- `Object Name`: This is the name of the object in the Google Cloud Storage bucket.\n",
+    "- `Path`:\n",
```
I think you need to adjust the spacing here, I see this: https://screenshot.googleplex.com/BTtCezJxPLfYTrq

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "\n",
+    "Widget input parameters include:\n",
+    "\n",
+    "- `Cloning`: _You cannot change the workspace ID after workspace creation._\n",
```
Fix this line, same issue as above

emmarogge replied:
fixed

wnojopra commented on first_hour_on_terra/workspace_resource_examples.ipynb:
```diff
@@ -0,0 +1,1552 @@
+    "Widget input parameters include:\n",
+    "\n",
+    "- `Cloning`: _You cannot change the workspace ID after workspace creation._\n",
+    "- `Bucket Name`: Must be a string that [meets requirements](https://cloud.google.com/storage/docs/buckets#naming) and is globally unique. We suggest prepending the `<PROJECT_ID>` of the GCP project associated with your workspace, which can be found on the to a descriptive string (e.g., `<PROJECT_ID>_output_data`).\n",
```
Grammar at end of this sentence, same issue as above

emmarogge replied:
fixed

wnojopra reviewed (changes requested):
Nice work. I've pointed out some obvious errors in the referenced resources section. I haven't yet review the controlled resources section, but please go over that section as well and fix any similar issues there.

emmarogge committed:
Fix grammar errors

Fix various errors spotted by reviewer in Markdwon cells for the referenced resource section.

emmarogge committed:
Merge branch 'main' into emmarogge/resources_PR

emmarogge committed:
Add parameter descriptions for cloud env

Add descriptions of parameters for widget that creates a cloud environment.

emmarogge committed:
Further encapsulate widget functionality

For better encapsulation, extract widget similarities into two new classes. Build individual text and dropdown input widgets from these classes.

emmarogge commented:
[Preview of new widget layout & styling](https://screenshot.googleplex.com/3SGE2czLm3R8zz3) which is used for all of the widgets.

emmarogge committed:
Fix typo

emmarogge committed:
Remove scratch code from notebook

emmarogge committed:
Split sentence into two

emmarogge committed:
Fix awkward sentence

wnojopra reviewed (changes requested):
Matt mentioned in chat that adding reference bq dataset didn't emit any success or failure message for him. I'm noticing the same: 
![image](https://github.com/DataBiosphere/terra-axon-examples/assets/6879774/d6971e79-0e60-4e4d-91f2-75a7e0342074)

Please take a look. LMK if you're unable to reproduce.

emmarogge committed:
Replace --description flag in commands

I mistakenly removed the --description flag from each widget's Workbench CLI command. I have now replaced these flags and the widgets work as expected.

emmarogge commented:
> Please take a look.

I have now fixed the issue and tested to validate that I can no longer reproduce. I had accidentally removed the `--description` flag from the Workbench CLI commands due to sloppy find & replace. It is now fixed.

emmarogge committed:
Provide label for BigQuery dataset input widget

emmarogge committed:
Update button labels

For controlled resources, button should read , not .

emmarogge committed:
Merge in changes from main

Merge in changes from main branch. Move the resources notebook to the new directory, , where the other Getting Started notebooks reside.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "\n",
+    "Then, populate the input fields:\n",
+    "\n",
+    "- `Cloning`: Read about these options [in the Cloning Instruction section](#choosing-the-right-cloning-instruction).\n",
```
Lets have the order of these input fields match the order thats presented in the UI

emmarogge replied:
done!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "Then, populate the input fields:\n",
+    "\n",
+    "- `Cloning`: Read about these options [in the Cloning Instruction section](#choosing-the-right-cloning-instruction).\n",
+    "- `Project ID`: The Google Cloud Project ID of your dataset.\n",
```
The UI calls it `GCP Project ID`, lets have it match here.

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "        self.input_description = TextInputWidget(\n",
+    "            \"<REPO_DESCRIPTION>\", \"Description:\").get()\n",
+    "        self.input_name = TextInputWidget(\"<NAME>\", \"Name:\").get()\n",
+    "        self.input_path = TextInputWidget(\"<PATH>\", \"Path:\").get()\n",
```
Can the UI make it clear which that either path field OR (dataset id AND project ID) must be populated? As is it feels like I need to fill out all fields. Even minimally just some explanation text under `Please provide appropriate values in the input boxes.` would suffice.

emmarogge replied:
addressed!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 3. Add a BigQuery table\n",
```
Same comments about datasets from above apply to this section.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "- `Bucket Name`: Must be a string that [meets requirements](https://cloud.google.com/storage/docs/buckets#naming) and is globally unique. A recommended strategy is to preprend the `<PROJECT_ID>` of the GCP project associated with your workspace to a descriptive string (e.g., `<PROJECT_ID>_output_data`).\n",
+    "- `Description`: This value is shown in the Resources tab of your workspace's page in Verily Workbench and when the `terra resource list` command is run. The description you provide should add helpful context about the purpose and/or contents of your bucket.\n",
+    "- `Name`: This name can be referenced in Workbench CLI commands. The value should be brief and memorable while communicating the purpose and/or contents of the bucket (e.g. `<STUDY_NAME>-data-bucket`).\n",
+    "- `Object Name`: The name of the object in the Google Cloud Storage bucket (e.g. `my_data.csv`).\n",
```
Same comment here about making it more clear in the UI that not all fields are required.

emmarogge replied:
fixed.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "For each type of referenced resource supported by Enterprise Terra, this notebook provides\n",
+    "\n",
+    "1. A **widget** which runs a Terra CLI command to add a referenced resource.<br>simply fill in the inputs with your desired values and click the button.\n",
+    "1. A **video** showing how to create that type of resource in the Enterprise Terra UI.\n"
```
Do we have videos for this? If not, remove this line

emmarogge replied:
fixed.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "\n",
+    "1. A widget which runs a Terra CLI command to add a referenced resource.\n",
+    "   Simply run the cell to create teh widget, fill in the inputs with your desired values and click the button.\n",
+    "1. A video showing how to create that type of resource in the Enterprise Terra UI.\n"
```
Remove or add videos

emmarogge replied:
fixed

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+   "source": [
+    "### Add referenced resources\n",
+    "\n",
+    "[Referenced resources](https://terra-docs.api.verily.com/docs/getting_started/web_ui/#referenced-vs-workspace-controlled-resources) represent data and other elements in Enterprise Terra by pointing to a source that exists outside of the current workspace.\n",
```
There's a few stray references to Enterprise Terra that should be VWB in this section

emmarogge replied:
fixed

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "For each type of controlled resource supported in Enterprise Terra, this notebook provides:\n",
+    "\n",
+    "1. A widget which runs a Terra CLI command to add a referenced resource.\n",
+    "   Simply run the cell to create teh widget, fill in the inputs with your desired values and click the button.\n",
```
```suggestion
    "   Simply run the cell to create the widget, fill in the inputs with your desired values and click the button.\n",
```

emmarogge replied:
fixed

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+    "                self.input_description,\n",
+    "                self.input_lifecycle,\n",
+    "                self.storage_drop_down,\n",
+    "                self.input_name,\n",
```
This vbox is missing the button

emmarogge replied:
fixed

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1436 @@
+   },
+   "outputs": [],
+   "source": [
+    "class CreateGcpNotebookWidget(object):\n",
```
Running this cell I get `NameError: name 'widget_layout' is not defined`

emmarogge replied:
fixed.

emmarogge committed:
Updates due to reviewer feedback

- Replace use of Enterprise Terra with Verily Workbench
- Remove mention of videos
- Use GCP Project ID, not Project ID
- Add missing button
- Indicate in UI which input fields are required/optional
- Fix typos

emmarogge committed:
Encapsulate button as class

Encapsulate button as a class to remove repeated layout and styling code from each widget.

emmarogge committed:
Mark a few more input fields optional.

emmarogge committed:
Update a few straggler instances of project ID to GCP project ID

emmarogge committed:
Reorder parameter descriptions to match widget order

Reorder parameter desscriptions in Markdown to match the ordering in each widget. Provide additional detail in a few spots.

emmarogge committed:
Remove location flag from command where no location is given

emmarogge committed:
Reorder GCP NB creation widget Markdown

emmarogge committed:
Remove mention of machine type parameter

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1471 @@
+    "        self.label = widgets.Label(\n",
+    "            value=\"Please provide appropriate values in the input boxes.\")\n",
+    "        self.warning = widgets.Label(\n",
+    "            value = \"If you don't provide the path, you must provide both GCP Project ID and Dataset ID.\")\n",
```
```suggestion
    "            value = \"Provide either the path, or BOTH Dataset ID and GCP Project ID. All other fields are required.\")\n",
```

wnojopra replied:
This suggestion is more direct and brief, while also clarifying which fields are required. If you accept this change please repeat for the other similar warnings in this notebook.

emmarogge replied:
updated throughout!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1471 @@
+    "        self.input_lifecycle = TextInputWidget(\n",
+    "            \"<PATH_TO_JSON>\", \"Lifecycle (JSON) - OPTIONAL:\").get()\n",
+    "        self.input_location = TextInputWidget(\n",
+    "            \"<LOCATION>\", \"Location - OPTIONAL:\").get()\n",
```
```suggestion
    "            \"<AUTODELETE (DAYS)>\", \"Auto Delete (Days) - (optional):\").get()\n",
    "        self.input_bucket_name = TextInputWidget(\n",
    "            \"<BUCKET_NAME>\", \"Bucket Name:\").get()\n",
    "        self.cloning_drop_down = DropdownInputWidget(\n",
    "            ['COPY_NOTHING', 'COPY_DEFINITION', 'COPY_RESOURCE', 'COPY_REFERENCE'], 'COPY_REFERENCE', \"Cloning:\").get()\n",
    "        self.input_description = TextInputWidget(\n",
    "            \"<DESCRIPTION>\", \"Description:\").get()\n",
    "        self.input_name = TextInputWidget(\"<NAME>\", \"Name:\").get()\n",
    "        self.input_lifecycle = TextInputWidget(\n",
    "            \"<PATH_TO_JSON>\", \"Lifecycle (JSON) - (optional):\").get()\n",
    "        self.input_location = TextInputWidget(\n",
    "            \"<LOCATION>\", \"Location - (optional):\").get()\n",
```

wnojopra replied:
In this UI, we also use ALL CAPS for enums, so OPTIONAL kinda blended in with that

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1471 @@
+    "- `Cloning`: Read about these options [in the Cloning Instruction section](#choosing-the-right-cloning-instruction). Default is `COPY_REFERENCE`.\n",
+    "- `Description`: (_Optional_) Should describe the purpose of the cloud environment.\n",
+    "- `Instance ID`: (_Optional_) A unique name given to the cloud environment which cannot be changed later. We recommend not specifying the ID, in which case an auto-generated one based on your email address and the time will be used. This ID is visible in the Google Cloud console for debugging, not in the VWB UI.\n",
+    "- `Location`: (_Optional_) The [Google Cloud location]((https://cloud.google.com/vertex-ai/docs/general/locations#user-managed-notebooks-locations) of your cloud environment. Default is `us-central1-a`.\n",
```
I think you need to remove this Location line now, right?

emmarogge replied:
I wasn't sure whether it'd be worthwhile to add more conditional logic based on whether the user provides a location in the future. Since I haven't implemented that yet, I'll remove the parameter description for now.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -0,0 +1,1471 @@
+    "- `Instance ID`: (_Optional_) A unique name given to the cloud environment which cannot be changed later. We recommend not specifying the ID, in which case an auto-generated one based on your email address and the time will be used. This ID is visible in the Google Cloud console for debugging, not in the VWB UI.\n",
+    "- `Location`: (_Optional_) The [Google Cloud location]((https://cloud.google.com/vertex-ai/docs/general/locations#user-managed-notebooks-locations) of your cloud environment. Default is `us-central1-a`.\n",
+    "- `Name`: This name is displayed in the VWB UI Environments tab. The value should be brief and memorable while communicating the purpose of the cloud environment (e.g. `analysis-notebooks-cloud-env`).\n",
+    "- `Post-startup Script`: (_Optional_) Path to a Bash script that automatically runs after your cloud env fully boots up. The path must be a URL or Cloud Storage path, e.g. 'gs://path-to-file/file-name'. If no post-startup script is provided, the [default script](https://raw.githubusercontent.com/DataBiosphere/terra-workspace-manager/main/service/src/main/java/bio/terra/workspace/service/reso ...
```
```suggestion
    "- `Post-startup Script`: (_Optional_) Path to a Bash script that automatically runs after your cloud env fully boots up. The path must be a URL or Cloud Storage path, e.g. 'gs://path-to-file/file-name'. If no post-startup script is provided, the [default script](https://github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh) will be used, which is recommended in most cases.\n",
```

wnojopra replied:
Is there a particular reason to send them to the raw version of the script? I think it'd be nice to use the regular page which has syntax highlighting, and other context.

emmarogge replied:
No particular reason -- and you're right, it's a better experience to provide the other link. Updated.

wnojopra reviewed (approved):
Wonderful!! This is looking great. Thanks for sticking through with this and working on it. I've got a few more nits you to resolve but other than that LGTM

emmarogge committed:
Updates based on reviewer feedback

Updates include:
- Refactor UI warnings about required inputs
- Use , not  for field labels
- Remove location as a parameter for GCP notebook (cloud env)
- Replace link to raw version of post-startup script with nicer GitHub link

emmarogge committed:
Update straggler instances of OPTIONAL
